sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMemoryLeaseManager.scala

```diff
@@ -0,0 +1,347 @@
+  // be manipulated while holding the GpuMemoryLeaseManager lock. (so essentially just always
+  // grab the lock before doing anything. This includes state in TrackingForTask. To keep
+  // things simple all threads that are waiting for more memory will be notified when enough
+  // memory is available to handle the next thread, but only the threads associated with that
+  // task will stay awake and continue processing. All of the others will wait/block again.
+  class TrackingForTask {
+    var totalUsed: Long = 0
+    var waitingThreadCount: Long = 0
+    var canWakeUp: Boolean = false
+  }
+
+  // The total amount of memory available to be leased
+  private var memoryAvailable: Long = memoryForLease
+  // The current active tasks along with their state
+  private val activeTasks = new ConcurrentHashMap[Long, TrackingForTask]
+  // The list of tasks in priority order to be fulfilled. Currently this is FIFO order
+  private val taskPriorityQueue = new util.LinkedList[TrackingForTask]()
+  private var shuttingDown: Boolean = false
+
+  def shutdown(): Unit = synchronized {
+    shuttingDown = true
+    if (!activeTasks.isEmpty) {
+      logDebug(s"shutting down with ${activeTasks.size} tasks still registered")
+      notifyAll()
+    }
+  }
+
+  /**
+   * Wake up any tasks that can now run. This should be called any time that memory is returned
+   * and memoryAvailable increases. It must be called while the GpuMemoryLeaseManager lock is
+   * held. That is why it is a private method and should never be made public.
+   */
+  private def wakeTasksThatCanRun(): Unit = {
+    var anyWoke = false
+    var keepGoing = !taskPriorityQueue.isEmpty
+    while (keepGoing) {
+      val head = taskPriorityQueue.peek()
+      if (head.totalUsed <= memoryAvailable) {
+        taskPriorityQueue.pop()
+        memoryAvailable -= head.totalUsed
+        head.canWakeUp = true
+        keepGoing = !taskPriorityQueue.isEmpty
+        anyWoke = true
+      } else {
+        keepGoing = false
+      }
+    }
+    if (anyWoke) {
+      notifyAll()
+    }
+  }
+
+  /**
+   * Request a new lease. This can throw if the requiredAmount cannot be fulfilled.
+   * @param tc the task the lease is for
+   * @param optionalAmount the amount optionally needed (amount might be lowered if needed)
+   * @param requiredAmount the amount that is required to work (not compatible with
+   *                       nonBlocking = true)
+   * @param nonBlocking can the request block to get the requested resources, or not.
+   * @return the lease.
+   */
+  def requestLease(tc: TaskContext,
+      optionalAmount: Long,
+      requiredAmount: Long,
+      nonBlocking: Boolean): MemoryLease = synchronized {
+    val taskAttemptId = tc.taskAttemptId()
+    if (nonBlocking && requiredAmount > 0) {
+      throw new IllegalArgumentException("Non-blocking requests with a required minimum amount " +
+          "of memory are not supported, because they are too likely to fail randomly")
+    }
```

Comment:
```suggestion
    require(nonBlocking && requiredAmount > 0,
      "Non-blocking requests with a required minimum amount " +
          "of memory are not supported, because they are too likely to fail randomly")
```